hybrid/flow.py

```diff
@@ -623,34 +718,128 @@ def __repr__(self):
-        """Execute one blocking iteration of an instantiated :class:`Loop`."""
-        last = state
-        last_quality = self.key(last)
+        iterno = 0
         cnt = self.convergence
+        input_state = state
 
-        for iterno in range(self.max_iter):
-            state = self.runnable.run(state, executor=immediate_executor).result()
-            state_quality = self.key(state)
+        while True:
+            output_state = self.runnable.run(input_state, executor=immediate_executor).result()
 
-            logger.info("{name} Iteration(iterno={iterno}, best_state_quality={key})".format(
-                name=self.name, iterno=iterno, key=state_quality))
+            iterno, cnt, input_state = self.iteration_update(iterno, cnt, input_state, output_state)
 
-            if state_quality == last_quality:
-                cnt -= 1
-            else:
-                cnt = self.convergence
+            if self.max_iter is not None and iterno >= self.max_iter:
+                break
             if cnt <= 0:
                 break
 
-            last_quality = state_quality
+        return output_state
 
-        return state
+
+class LoopN(LoopUntilNoImprovement):
+    """Iterate `runnable` exactly `n` times, using each output as input in the
+    next iteration.
+    """
+
+    def __init__(self, runnable, n):
+        super(LoopN, self).__init__(
+            runnable=runnable, max_iter=n, convergence=n, key=lambda _: 0)
 
 
+class Loop(LoopUntilNoImprovement):
+    pass
+
 SimpleIterator = Loop
 
 
+class LoopWhileNoImprovement(LoopUntilNoImprovement):
+    """Iterates `runnable` until a state quality metric, defined by the `key`
+    function, shows no improvement for at least `max_tries` number of
+    iterations.
+
+    Args:
+        runnable (:class:`Runnable`):
+            A runnable that's looped over.
+
+        max_tries (int, optional, default=10):
+            Maximum number of times the `runnable` is run for the **same** input
+            state. On each improvement, the better state is used for the next
+            input state, and tries counter is reset.
+
+        key (callable/str):
+            Best state is judged according to a metric defined with a `key`.
+            `key` can be a `callable` with a signature::
+
+                key :: (State s, Ord k) => s -> k
+
+            or a string holding a key name/path to be extracted from the input
+            state with `operator.attrgetter` method.
+
+            By default, `key == operator.attrgetter('samples.first.energy')`,
+            thus favoring states containing a sample with the minimal energy.
+
+    """
+
+    def __init__(self, runnable, max_tries=10, key=None):
+        super(LoopWhileNoImprovement, self).__init__(
+            runnable=runnable, max_iter=None, convergence=max_tries, key=key)
+
+    def iteration_update(self, iterno, cnt, input_state, output_state):
+        """Implement "no-improvement count-down" behavior:
```

Comment:
Why not just use the two bullets as sentences without this introduction statement?

Reply:
Adds a bit of structure. I can't start a paragraph with a bullet point. :)

Reply:
@JoelPasvolsky Thanks for all the suggestions and improvements!